framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,605 @@
+     *     for parameters, but must be correct for fields.
+     * @param className the name of the enclosing class. This is only used for printing the type of
+     *     an explicit receiver parameter (i.e., a parameter named "this").
+     * @return a String suitable to print in a stub file
+     */
+    private String formatAField(AField aField, String fieldName, String className) {
+        StringBuilder result = new StringBuilder();
+        String basetype;
+        if ("this".equals(fieldName)) {
+            basetype = className;
+        } else if (basetypes.containsKey(aField.type.description.toString())) {
+            basetype = basetypes.get(aField.type.description.toString()).toString();
+        } else {
+            throw new BugInCF(
+                    "SceneToStubWriter: could not find the base type for this variable declaration: "
+                            + fieldName);
+        }
+
+        if (basetype.contains("[")) {
+            String component = basetype.substring(0, basetype.lastIndexOf('['));
+            result.append(formatArrayComponentTypeAnnotation(aField.type));
+            result.append(component);
+            result.append(" ");
+            basetype = "[]";
+        }
+        result.append(formatAnnotations(aField.type.tlAnnotationsHere));
+        result.append(basetype);
+        result.append(' ');
+        result.append(fieldName);
+        return result.toString();
+    }
+
+    /**
+     * Writes out an import statement for each annotation used in an {@link AScene}.
+     *
+     * <p>{@code DefCollector} is a facility in the Annotation File Utilities for determining which
+     * annotations are used in a given AScene. Here, we use that construct to write out the proper
+     * import statements into a stub file.
+     */
+    private static class ImportDefCollector extends DefCollector {
+
+        /** The writer onto which to write the import statements. */
+        private final PrintWriter printWriter;
+
+        /**
+         * Constructs a new ImportDefCollector, which will run on the given AScene when its {@code
+         * visit} method is called.
+         *
+         * @param scene the scene whose imported annotations should be printed
+         * @param printWriter the writer onto which to write the import statements
+         * @throws DefException if the DefCollector does not succeed
+         */
+        ImportDefCollector(AScene scene, PrintWriter printWriter) throws DefException {
+            super(scene);
+            this.printWriter = printWriter;
+        }
+
+        /**
+         * Write an import statement for a given AnnotationDef. This is only called once per
+         * annotation used in the scene.
+         *
+         * @param d the annotation definition to print an import for
+         */
+        @Override
+        protected void visitAnnotationDef(AnnotationDef d) {
+            if (!d.name.contains("+")) {
```

Comment:
I suggest abstracting this out.

Reply:
I added a comment explaining why this was done, but it seems silly to abstract out, since it's only checked once. Is there a particular reason you think it should be abstracted?